presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -1334,4 +1361,272 @@ private PartitioningVariableAssignments(List<VariableReferenceExpression> variab
             return constants;
         }
     }
+
+    private static class IsQueryValueNodeVisitor
+            extends InternalPlanVisitor<Boolean, Void>
+    {
+        @Override
+        public Boolean visitPlan(PlanNode node, Void context)
+        {
+            ImmutableList.Builder<PlanNode> leafNodeBuilder = ImmutableList.builder();
+            findLeafNodes(node, leafNodeBuilder);
+            boolean leafIsValueNode = true;
+            for (PlanNode planNode : leafNodeBuilder.build()) {
+                leafIsValueNode = leafIsValueNode && planNode.accept(this, context);
+            }
+            return leafIsValueNode;
+        }
+
+        private void findLeafNodes(PlanNode node, ImmutableList.Builder<PlanNode> leafNodeBuilder)
+        {
+            if (node.getSources() == null || node.getSources().size() == 0) {
+                leafNodeBuilder.add(node);
+            }
+            else {
+                for (PlanNode source : node.getSources()) {
+                    findLeafNodes(source, leafNodeBuilder);
+                }
+            }
+        }
+
+        @Override
+        public Boolean visitValues(ValuesNode node, Void context)
+        {
+            return true;
```

Comment:
It feels like you only need to override 3 methods:

- `visitValues` and make it return `true`

Then the default `visitPlan` should only piggyback on the result from child nodes. Only a single special case is needed there- if there's no child nodes (that indicates that a node is a "leaf" node and `false` should be returned as queries that have other leafs that `values` should not be allowed to run on coordinator)

Reply:
fix